text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+        world.create(State.from(new_asset));
+
+        Ok(())
+     }
+```
+
+Access to the ledger is achieved via the `Ledger::access_ledger()`. Access then to the 'world-state' or default collection is via `getCollection(...)`
+
+A key is needed and the `State::make_composite()` provides a helper function to create this key. `collection.create(state)` stores the state.  
+
+#### Update and Create
+
+Previously a stub API of `putState` was used; issues have been raised referring to what is called 'blind writes'. Namely, if a key is **put** without first being **get**  then the key will only be in the write set. This can have the effect that if two transactions do this the last one wins.
+
+A solution presented in one of the [JIRA Issues](https://jira.hyperledger.org/browse/FAB-10480) was the create and update semantics.
+
+- Create: if the key already exists fail, only put if the key does not exist.
+- Update: if the key already exists put the state, if it doesn't fail. 
+
+### Retrieval and Deleting
+
+Retrieval and Deleting of state follow similar patterns. The key lines are
+
+```rust
+    // get the state
+    String key = State::make_composite(...);
+    let asset: MyAsset = collection.retrieve(key).into_datatype();
+    // ....
+    let asset: MyAsset = collection.delete(key).into_datatype();
+
+```
+
+### State-Based Endorsement
+
+Endorsement Policies can be defined at the Contract, State, and (in v2.0) the collection level. Control of the State endorsement is by the contract code invoking APIs on the State object.  (details of the syntax are defined in the main docs)[https://hyperledger-fabric.readthedocs.io/en/master/endorsement-policies.html#endorsement-policy-syntax]
+
+
+```rust
+    // Create three principals
+    let p1 = Principal::new("Org1", Role.MEMBER);
+    let p2 = Principal::new("Org2", Role.MEMBER);
+    let p3 = Principal::new("Org3", Role.MEMBER);
+
+    // create an Endorsment based on the required logic
+    let sbe1 = Endorsement::build(
+            Expression::and(
+                    Expression::or(p1,p2),
+                    p3
+                    )
+            );
+    assetState.set_endorsement(sbe1);
+    
+    // or the textual version can be parsed
+    let sbe2 = Endorsement::build("AND( OR ('Org1.member','Org2.member'), 'Org3.member' )");
+    assetState.set_endorsement(sbe2);
+        
+```
+
+### Query
+
+There are two broad approaches to querying data. Either by passing a query to be parsed by CouchDB (also called Rich Query), or by using the State keys. Composite keys form a hierarchy that can be used for querying a range of states.
+
+To unify the different approaches, the API has a query method taking a configuration of the query to be performed, that returns an iterable object of results.  This iterable will contain the results, and be iterated with the appropriate language idioms. Control of pagination and the request for more data is controlled within this iterable.
```

Comment:
> To unify the different approaches, the API has a query method taking a configuration of the query to be performed, that returns an iterable object of results.  This iterable will contain the results, and be iterated with the appropriate language idioms.

I don't think you can really unify them, as the developer should be aware of what which of the two query types is going to be run. 
The reason is that rich queries do not get re-executed during commit, and therefore you cannot combine writes with rich queries.

Reply:
The developer is aware of the type of query they are performing: precise syntax might vary but...
```
collection.get_states(KeyQueryHandler::Range("Car001","Car002"));
collection.get_states(RichQueryHandler::Range("select * where car_age>2 and car_price>20000"));
```